core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -123,13 +123,13 @@ private static void addClassEntities(ImmutableSet.Builder<String> entities, Clas
         if (!isPublic(clazz.getModifiers())) {
             return;
         }
+        if (!includeDeprecated && clazz.isAnnotationPresent(Deprecated.class)) {
+            return;
+        }
```

Comment:
We should still execute `for (Class<?> nestedClass : clazz.getDeclaredClasses())`.
i.e. i think we should require each nested class to be _also_ deprecated before all are removed.

wdyt?

Reply:
Not sure. It feels ok to assume that if class is deprecated then all "inner" stuff (methods, nested classes) is also deprecated.

Reply:
Depends how this is supported by an IDE.

In my IntelliJ the following code has no deprecation warning, even at the import line.

```java
import io.trino.t.ADeprecatedClass.NonDeprecatedInnerClass;

public class CallSite
{
    public static void main(String[] args)
    {
        NonDeprecatedInnerClass x = new NonDeprecatedInnerClass();
        System.out.println(x.toString());
    }
}
```

i see a deprecation only if i write this like

```java
ADeprecatedClass.NonDeprecatedInnerClass x = new ADeprecatedClass.NonDeprecatedInnerClass();
```

since this is SPI, we shouldn't assume calling code to be written with any particular code style, and we should assume a developer may be using IntelliJ.

Reply:
@findepi this makes sense to me. I changed so that it still executes the declared classes loop even if the outer class is deprecated.